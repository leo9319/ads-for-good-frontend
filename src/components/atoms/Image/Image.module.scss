.imgWrapper {
  width: var(--imgWidth) !important;
  height: var(--imgHeight) !important;
  min-height: var(--imgMinHeight);
  min-width: var(--imgMinWidth);
  max-height: var(--imgMaxHeight);
  max-width: var(--imgMaxWidth);

  &.roundLight {
    border-radius: 10px !important;
  }
  &.roundMedium {
    border-radius: 20px !important;
  }
  &.roundFull {
    border-radius: 50% !important;
  }
}
