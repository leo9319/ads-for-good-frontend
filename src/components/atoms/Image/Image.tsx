import React, { useEffect, useState } from 'react';
import { Skeleton } from '@radix-ui/themes';
import { classNames, getModuleClasses } from '@utils/common/classNames';
import { getStylesAsCssProperties } from '@utils/common/styles';
import { Link, LinkProps } from '@radix-styles/atoms/Link';
import AspectRatio, {
  AspectRatioType,
} from '@radix-styles/atoms/AspectRatio/AspectRatio';

import styles from './Image.module.scss';

export interface ImageProps extends React.HTMLAttributes<HTMLImageElement> {
  /**
   * Source for the image, if the image source is not there then the image component will not load
   * @property
   */
  src: string | undefined;
  /**
   * Alternate text for the image
   * @property
   */
  alt: string | undefined;
  /**
   * Fallback image Source for the image, this is will work if actual image is broken
   * @property
   */
  fallbackSrc: string | undefined;
  /**
   * Class name for the image
   * @property
   */
  className?: string;
  /**
   * Width for the image
   * @property
   */
  width?: string;
  /**
   * Height for the image
   * @property
   */
  height?: string;
  /**
   * Minimum Height for the image
   * @property
   */
  minHeight?: string;
  /**
   * Minimum Width for the image
   * @property
   */
  minWidth?: string;
  /**
   * Maximum Height for the image
   * @property
   */
  maxHeight?: string;
  /**
   * Maximum Width for the image
   * @property
   */
  maxWidth?: string;
  /**
   * Link for the image
   * @property
   */
  link?: LinkProps;
  /**
   * Icon for the image, this will be used if the image source is not there
   * @property
   */
  icon?: string;
  /**
   * The aspect ratio to maintain (width / height)
   * For example, 16 / 9 or 4 / 3
   */
  aspectRatio?: AspectRatioType;
}

export const Image = ({
  src,
  alt,
  fallbackSrc,
  className,
  width,
  height,
  minHeight,
  minWidth,
  maxHeight,
  maxWidth,
  link,
  icon,
  aspectRatio,
  ...props
}: ImageProps) => {
  const [imgSrc, setImgSrc] = useState<string | undefined>(src);
  const [loading, setLoading] = useState(false);

  const iconClassName = icon ? `icon ${icon}` : '';
  className = classNames(iconClassName, className);
  const classes = getModuleClasses(className?.trim(), styles);
  const linkProps =
    typeof link === 'object' && !(link instanceof Array)
      ? link
      : { ignoreHref: true };

  const handleError = () => {
    setImgSrc(fallbackSrc);
    setLoading(false);
  };

  const onLoad = () => {
    setLoading(false);
  };

  useEffect(() => {
    setImgSrc(src);
  }, [src]);

  useEffect(() => {
    setLoading(true);
  }, [imgSrc]);

  if (!src && !icon) {
    return null;
  }

  const hasRatio = aspectRatio && aspectRatio !== 'none';

  return (
    <Link {...linkProps}>
      <Skeleton loading={loading}>
        <AspectRatio ratio={aspectRatio}>
          <img
            src={imgSrc}
            alt={alt}
            onError={handleError}
            onLoad={onLoad}
            className={classNames(styles.imgWrapper, classes)}
            style={getStylesAsCssProperties({
              '--imgWidth': hasRatio ? '100%' : width,
              '--imgHeight': hasRatio ? '100%' : height,
              '--imgMinHeight': minHeight,
              '--imgMinWidth': minWidth,
              '--imgMaxHeight': maxHeight,
              '--imgMaxWidth': maxWidth,
            })}
            {...props}
          />
        </AspectRatio>
      </Skeleton>
    </Link>
  );
};

export default Image;
