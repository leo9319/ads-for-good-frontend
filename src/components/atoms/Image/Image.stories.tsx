import type { Meta, StoryObj } from '@storybook/react';
import Image from './Image';
import FallbackImage from '@assets/images/FallbackImage.svg';

const meta = {
  title: 'Library/Atoms/Image',
  component: Image,
  parameters: {
    layout: 'fullscreen',
  },
  tags: ['autodocs'],
  argTypes: {
    src: {
      control: { type: 'text' },
    },
    alt: {
      control: { type: 'text' },
    },
    fallbackSrc: {
      control: { type: 'text' },
    },
    className: {
      control: { type: 'text' },
    },
    width: {
      control: { type: 'text' },
    },
    height: {
      control: { type: 'text' },
    },
    minHeight: {
      control: { type: 'text' },
    },
    minWidth: {
      control: { type: 'text' },
    },
    maxHeight: {
      control: { type: 'text' },
    },
    maxWidth: {
      control: { type: 'text' },
    },
    link: {
      control: { type: 'object' },
    },
    aspectRatio: {
      control: {
        type: 'select',
      },
      options: ['1:1', '2:3', '3:2', '3:4', '4:3', '9:16', '16:9', 'none'],
    },
  },
  args: {
    link: { ignoreHref: true },
  },
} satisfies Meta<typeof Image>;

export default meta;
type Story = StoryObj<typeof meta>;

export const image: Story = {
  args: {
    src: 'https://www.gstatic.com/webp/gallery/1.jpg',
    alt: 'dummy alt',
    fallbackSrc: FallbackImage,
    className: '',
    width: '100px',
    height: '100px',
    minHeight: '100px',
    minWidth: '100px',
    maxHeight: '100px',
    maxWidth: '100px',
  },
};

export const imageWithLink: Story = {
  args: {
    src: 'https://www.gstatic.com/webp/gallery/1.jpg',
    alt: 'dummy alt',
    fallbackSrc: FallbackImage,
    className: '',
    width: '100px',
    height: '100px',
    minHeight: '100px',
    minWidth: '100px',
    maxHeight: '100px',
    maxWidth: '100px',
    link: {
      href: 'https://www.google.com/',
      target: '_blank',
    },
  },
};
