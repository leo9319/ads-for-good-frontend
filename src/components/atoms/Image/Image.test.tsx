import React from 'react';
import Image from './Image';
import { render } from '@testing-library/react';

describe('Image', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <Image
        src="https://images.unsplash.com/photo-1479030160180-b1860951d696?&auto=format&fit=crop&w=1200&q=80"
        alt="Alt text"
        fallbackSrc=""
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
