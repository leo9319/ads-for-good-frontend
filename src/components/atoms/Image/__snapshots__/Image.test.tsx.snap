// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should render successfully 1`] = `
<body>
  <div>
    <img
      alt="Alt text"
      class="imgWrapper"
      src="https://images.unsplash.com/photo-1479030160180-b1860951d696?&auto=format&fit=crop&w=1200&q=80"
    />
  </div>
</body>
`;
